torch/_inductor/fx_passes/b2b_gemm.py

```diff
@@ -154,118 +298,339 @@ def cdiv(x: int, y: int):
-    CallFunction(
-        torch.ops.aten.mm, CallFunction(torch.ops.aten.mm, Arg(), Arg()), Arg()
-    ),
-    extra_check=can_apply_b2b_gemm,
+    CallFunction(torch.ops.aten.mm, Arg(), Arg()),
     pass_dict=B2B_GEMM_PASS,
 )
 def b2b_gemm(
-    match: Match, mat1: torch.fx.Node, mat2: torch.fx.Node, mat3: torch.fx.Node
+    match: Match, mat1: torch.fx.Node, mat2: torch.fx.Node
 ) -> None:
+    # match.args: list[torch.fx.Node]
+
+    def is_pointwise_node(node: torch.fx.Node) -> bool:
+        return (
+            node.op == "call_function" and
+            isinstance(node.target, torch._ops.OpOverload) and
+            (torch.Tag.pointwise in node.target.tags)
+        )
+    
+    def is_mm(node: torch.fx.Node) -> bool:
+        return (
+            node.op == "call_function" and
+            node.target == torch.ops.aten.mm.default
+        )
+
+    # the inner MM
+    inner_mm = match.nodes[-1]
+
+    # find the (candidate) outer MM
+    # In a real (A @ f(B @ C)), every path starting from (B @ C) must reach (A @ _).
+    outer_mm = None
+    node = inner_mm
+    while len(node.users) > 0:
+        node = next(iter(node.users))
+        if is_mm(node):
+            outer_mm = node
+            break
+        elif is_pointwise_node(node):
+            continue
+        else:
+            break
+    if not outer_mm:
+        return
+
+    # find the unique input node for outer_mm representing f(B @ C) in (A @ f(B @ C))
+    # we call it the "f_node"
+    # when the pattern is simply (A @ (B @ C)), f_node is just inner_mm
+    f_node = inner_mm
+    while not (next(iter(f_node.users)) is outer_mm):
+        f_node = next(iter(f_node.users))
+
+    def all_reach_via(
+        src: torch.fx.Node,
+        dst: torch.fx.Node,
+        check_intermediate: Callable[[torch.fx.Node], bool],
+        get_next: Callable[[torch.fx.Node], Iterable[torch.fx.Node]],
+    ) -> bool:
+        '''
+        checks whether every path from src reaches dst via check_intermediate nodes
+        where the direction is specified by get_next
+        '''
+        visited: Dict[torch.fx.Node, bool] = {}
+        def dfs(node: torch.fx.Node) -> bool:
+            if node in visited:
+                return visited[node]
+            if node is dst:
+                ret = True
+            elif (node is src) or check_intermediate(node):
+                ret = True
+                for next_node in get_next(node):
+                    val = dfs(next_node)
+                    ret = ret and val
+            else:
+                ret = False
+            visited[node] = ret
+            return ret
+        
+        return dfs(src)
+
+    # checker inner_mm and f_node are connected by pointwise nodes with no outside input_nodes / users
+    if not all_reach_via(inner_mm, f_node, is_pointwise_node, lambda node: iter(node.users)):
+        return
+    if not all_reach_via(f_node, inner_mm, is_pointwise_node, lambda node: node.all_input_nodes):
+        return
```

Comment:
do we need both of these ?

Reply:
The above loop first tries to find a "candidate" outer_mm and also f_node, while the two checks here are for final confirmation.

If we don't know the outer_mm, when doing dfs we will need to confirm we are reaching "the same" mm via every path, which (I feel) requires more bookkeeping efforts.

Reply:
I understand we are finding the `inner_mm`, but I dont think we need to iterate from inner_mm -> f_node and from f_node -> inner_mm

Reply:
I think one iteration is for .users while the other iteration is for .input_nodes.

If we just check inner_mm -> f_node via .users then there might be other inputs to the subgraph, which isn't tracked by the current subgraph construction.

Similarly if we just check f_node -> inner_mm via .input_nodes then there might be other users of the subgraph, in which case we cannot erase the subgraph.

Reply:
If at the end, there is an indeg, then that checks if there are other inputs to the subgraph

Reply:
I updated the algorithm accordingly. Now it only requires one iteration and all unit tests still pass.